File: app/schemas/traces.py
"""Schema models for pipeline trace responses."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.pipelines.definition import PipelineDefinition
from app.schemas.base import DateTimeConfigMixin
from app.schemas.enums import PipelineIOType, PipelineKind, PipelineRunStatus


class PipelineRunRead(DateTimeConfigMixin, BaseModel):
    """Pipeline run details returned to clients.

    Built via `PipelineRunRead.model_validate(run, from_attributes=True)` --
    every declared field is a plain column on `models.PipelineRun`.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    pipeline_id: UUID
    pipeline_version_id: UUID | None
    pipeline_version: int | None
    kind: PipelineKind
    user_id: UUID
    collection_id: UUID
    status: PipelineRunStatus
    error_message: str | None
    warnings: list[str] = Field(default_factory=list)
    started_at: datetime
    completed_at: datetime | None


class PipelineNodeSummaryValueRead(BaseModel):
    """Summary value for a pipeline node input/output."""

    label: str
    value: Any
    kind: Literal["json", "text", "embedding", "items", "ranking"] = "json"


class PipelineNodeSummaryRead(BaseModel):
    """Summary of key inputs and outputs for a node."""

    inputs: list[PipelineNodeSummaryValueRead] = Field(default_factory=list)
    outputs: list[PipelineNodeSummaryValueRead] = Field(default_factory=list)


class PipelineNodeRunRead(DateTimeConfigMixin, BaseModel):
    """Node execution details returned to clients.

    Built via `PipelineNodeRunRead.model_validate(node_run, from_attributes=True)`
    -- `summary` is a plain `dict` column on `models.PipelineNodeRun`; nested
    validation into `PipelineNodeSummaryRead` happens the same way it would
    for a dict passed to `model_validate` directly.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID
    node_id: str
    node_type: str
    node_name: str
    sequence_index: int
    status: PipelineRunStatus
    error_message: str | None
    started_at: datetime
    completed_at: datetime | None
    duration_ms: float | None
    summary: PipelineNodeSummaryRead


class PipelineNodeIORead(DateTimeConfigMixin, BaseModel):
    """Node input/output payloads returned to clients.

    Built via `PipelineNodeIORead.model_validate(io_record, from_attributes=True)`.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID
    node_run_id: UUID
    node_id: str
    io_type: PipelineIOType
    port: str
    payload: dict[str, Any]


class PipelineTraceResponse(BaseModel):
    """Pipeline trace response payload."""

    run: PipelineRunRead
    definition: PipelineDefinition
    node_runs: list[PipelineNodeRunRead]
    node_io: list[PipelineNodeIORead]


class TraceOriginRead(BaseModel):
    """Where a retrieved chunk came from: its source document and the
    ingestion run that put it into the index."""

    document_id: UUID
    document_name: str | None
    chunk_id: str | None
    trace: PipelineTraceResponse


class FocusedItemRead(BaseModel):
    """The concrete chunk behind a focused trace item.

    Trace identity lists carry only ids and scores; this is the live lookup
    that gives the focused result its text and document context. `status` is
    "missing" when the chunk id no longer resolves (deleted or re-ingested
    content) -- the journey still renders from the recorded lists.
    """

    id: str
    status: Literal["resolved", "missing"]
    text: str | None = None
    document_id: UUID | None = None
    filename: str | None = None
    chunk_index: int | None = None
    chunk_count: int | None = None


class DocumentTraceResponse(BaseModel):
    """An ingestion trace focused on one of the document's chunks.

    The chunk-focused variant of the plain document trace: same run payload,
    plus the live-resolved chunk behind the focused id.
    """

    trace: PipelineTraceResponse
    focused_item: FocusedItemRead | None = None
    context_items: list[FocusedItemRead] = Field(default_factory=list)


class EndToEndTraceResponse(BaseModel):
    """A retrieval trace joined with the origin ingestion trace.

    `origin` is None when the source document (or its ingestion run) can't be
    resolved -- the retrieval trace still stands on its own. `focused_item`
    is populated whenever a chunk id was requested, resolved or not.
    """

    retrieval: PipelineTraceResponse
    origin: TraceOriginRead | None = None
    focused_item: FocusedItemRead | None = None
    context_items: list[FocusedItemRead] = Field(default_factory=list)

File: frontend/src/lib/types/traces.ts
import type { UUID } from "@/lib/types/common";
import type {
  PipelineDefinition,
  PipelineIOType,
  PipelineKind,
  PipelineRunStatus,
} from "@/lib/types/pipelines";

export interface PipelineRunTrace {
  id: UUID;
  pipeline_id: UUID;
  pipeline_version_id?: UUID | null;
  pipeline_version?: number | null;
  kind: PipelineKind;
  user_id: UUID;
  collection_id: UUID;
  status: PipelineRunStatus;
  error_message?: string | null;
  warnings: string[];
  started_at: string;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PipelineNodeRunTrace {
  id: UUID;
  run_id: UUID;
  node_id: string;
  node_type: string;
  node_name: string;
  sequence_index: number;
  status: PipelineRunStatus;
  error_message?: string | null;
  started_at: string;
  completed_at?: string | null;
  duration_ms?: number | null;
  summary: PipelineNodeSummary;
  created_at: string;
  updated_at: string;
}

export interface PipelineNodeSummaryValue {
  label: string;
  value: unknown;
  kind?: "json" | "text" | "embedding" | "items" | "ranking";
}

export interface ItemRef {
  id: string;
  score?: number | null;
}

export interface ItemListTrace {
  kind: "chunks" | "matches";
  items: ItemRef[];
}

export interface RankingSourceEvidence {
  source_index: number;
  rank?: number | null;
  score?: number | null;
  score_label?: string | null;
  weight?: number | null;
  contribution?: number | null;
}

export interface RankingResultEvidence {
  id: string;
  rank: number;
  score?: number | null;
  sources: RankingSourceEvidence[];
}

export interface RankingEvidence {
  method: string;
  score_label?: string | null;
  formula?: string | null;
  results: RankingResultEvidence[];
}

export interface PipelineNodeSummary {
  inputs: PipelineNodeSummaryValue[];
  outputs: PipelineNodeSummaryValue[];
}

export interface PipelineNodeIOTrace {
  id: UUID;
  run_id: UUID;
  node_run_id: UUID;
  node_id: string;
  io_type: PipelineIOType;
  port: string;
  payload: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface PipelineTraceResponse {
  run: PipelineRunTrace;
  definition: PipelineDefinition;
  node_runs: PipelineNodeRunTrace[];
  node_io: PipelineNodeIOTrace[];
}

/** `TraceOriginRead` — the source document + ingestion trace for a chunk. */
export interface TraceOrigin {
  document_id: UUID;
  document_name?: string | null;
  chunk_id?: string | null;
  trace: PipelineTraceResponse;
}

/** `FocusedItemRead` — the concrete chunk behind a focused trace item. */
export interface TraceFocusedItem {
  id: string;
  status: "resolved" | "missing";
  text?: string | null;
  document_id?: UUID | null;
  filename?: string | null;
  chunk_index?: number | null;
  chunk_count?: number | null;
}

/** `DocumentTraceResponse` — ingestion trace with one chunk resolved for focus. */
export interface DocumentTrace {
  trace: PipelineTraceResponse;
  focused_item?: TraceFocusedItem | null;
  context_items: TraceFocusedItem[];
}

/** `EndToEndTraceResponse` — retrieval trace joined with chunk origin. */
export interface EndToEndTrace {
  retrieval: PipelineTraceResponse;
  origin?: TraceOrigin | null;
  focused_item?: TraceFocusedItem | null;
  context_items: TraceFocusedItem[];
}
